test/test_pipeline/test_classification.py

```diff
@@ -822,3 +823,37 @@ def test_fit_instantiates_component(self):
                 y=np.array([1, 0, 1, 1])
             )
         del preprocessing_components._addons.components['CrashPreprocessor']
+
+    def test_check_is_fitted(self):
+        """Makes sure that instantiating and estimator, does not cause the object
+        to be treated as fitted. That is, when an estimator is created, it should
+        not be recognized as a fitted object. Only after it is fitted
+
+        Sklearn adds a fitted_ var to the estimator, nevertheless the check_is_fitted
+        widely checks for any key in vars() ending with _. This unit testing
+        enforces that check_is_fitted works.
+        """
+
+        all_combinations = list(itertools.product([True, False], repeat=4))
+        for sparse, multilabel, signed, multiclass, in all_combinations:
+            dataset_properties = {
+                'sparse': sparse,
+                'multilabel': multilabel,
+                'multiclass': multiclass,
+                'signed': signed,
+            }
+            auto = SimpleClassificationPipeline(
+                random_state=1,
+                dataset_properties=dataset_properties,
+            )
+            cs = auto.get_hyperparameter_search_space()
+            config = cs.sample_configuration()
+
+            # Set hyperparameters takes a given config and translate
+            # a config to an actual implementation
+            auto.set_hyperparameters(config)
+
+            for name, step in auto.named_steps.items():
+                with self.assertRaisesRegex(sklearn.exceptions.NotFittedError,
+                                            "instance is not fitted yet"):
+                    check_is_fitted(step)
```

Comment:
Would it make sense to also call this on the full pipeline? So `check_is_fitted(auto)`? 

Also, would it make sense to actually fit the pipeline and check afterwards that it was fitted? Or maybe add that to some other unit test? As the issue was specifically about cloning, I think it would also make sense to fit, clone and then check if the estimator is still fitted, right?

Reply:
Thank you, this makes a lot of sense and I added a push with this:

- Move the checking for unfitted components to the configuration test, this way we can also check post fit (as all the try/catch blocks are here)
- We have a cloning check, and it did not trigger the problem, because I mistakenly assumed self.models should be clonable objects, yet in the case of cross validation this attribute is not supposed to be used. I added a clarification note for future readers.